File: vlm_op_profiler/cli/vlm_op_profiler.cpp
// vlm_op_profiler.cpp
//
// Process wrapper around llama-mtmd-cli.
//
// What it does:
//   1. Parses profiler-specific flags (--out-dir, --steps, etc.).
//   2. Locates libbackend_stats.so and llama-mtmd-cli relative to itself.
//   3. Creates the output directory.
//   4. Sets LD_PRELOAD / DYLD_INSERT_LIBRARIES so libbackend_stats is injected.
//   5. Sets PROFSTATS_OUT_DIR and PROFSTATS_MAX_STEPS for the interceptor.
//   6. exec's llama-mtmd-cli with the remaining (model / image / prompt) args.
//
// The profiler-specific flags are consumed here; everything else is forwarded
// to llama-mtmd-cli unchanged.

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>

#include <errno.h>
#include <sys/stat.h>
#include <unistd.h>

#ifdef __linux__
#  include <limits.h>  // PATH_MAX
#endif
#ifdef __APPLE__
#  include <mach-o/dyld.h>
#  include <limits.h>
#endif

// ---------------------------------------------------------------------------
// Version
// ---------------------------------------------------------------------------
static constexpr const char * VERSION = "0.1.0-dev";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static bool file_exists(const std::string & p) {
    struct stat st;
    return stat(p.c_str(), &st) == 0;
}

static bool is_executable(const std::string & p) {
    return access(p.c_str(), X_OK) == 0;
}

// Return the directory containing our own executable, with trailing '/'.
static std::string self_dir() {
    char buf[4096] = {};

#ifdef __linux__
    ssize_t len = readlink("/proc/self/exe", buf, sizeof(buf) - 1);
    if (len > 0) {
        buf[len] = '\0';
        char * slash = strrchr(buf, '/');
        if (slash) { slash[1] = '\0'; return buf; }
    }
#elif defined(__APPLE__)
    uint32_t size = sizeof(buf);
    if (_NSGetExecutablePath(buf, &size) == 0) {
        char * slash = strrchr(buf, '/');
        if (slash) { slash[1] = '\0'; return buf; }
    }
#endif
    return "./";
}

// Search candidate paths for an executable; return first hit.
static std::string find_binary(const std::string & name,
                                const std::vector<std::string> & candidates) {
    for (const auto & c : candidates) {
        if (is_executable(c)) return c;
    }
    // Fall back to PATH
    const char * path_env = std::getenv("PATH");
    if (path_env) {
        std::string path_str(path_env);
        size_t start = 0;
        while (start < path_str.size()) {
            size_t end = path_str.find(':', start);
            if (end == std::string::npos) end = path_str.size();
            std::string candidate = path_str.substr(start, end - start) + "/" + name;
            if (is_executable(candidate)) return candidate;
            start = end + 1;
        }
    }
    return {};
}

// Recursively create directories (like mkdir -p).
static bool mkdirp(const std::string & path) {
    if (path.empty()) return false;
    struct stat st;
    if (stat(path.c_str(), &st) == 0) return true;  // already exists

    // Create parent first.
    size_t slash = path.rfind('/');
    if (slash != std::string::npos && slash > 0) {
        if (!mkdirp(path.substr(0, slash))) return false;
    }

    if (mkdir(path.c_str(), 0755) != 0 && errno != EEXIST) {
        fprintf(stderr, "vlm-op-profiler: mkdir(%s): %s\n",
                path.c_str(), strerror(errno));
        return false;
    }
    return true;
}

// Prepend our library to LD_PRELOAD (or DYLD_INSERT_LIBRARIES on macOS),
// preserving any pre-existing value.
static void prepend_preload(const std::string & lib_path) {
#ifdef __APPLE__
    const char * key = "DYLD_INSERT_LIBRARIES";
#else
    const char * key = "LD_PRELOAD";
#endif
    const char * existing = std::getenv(key);
    std::string value = lib_path;
    if (existing && *existing) value += std::string(":") + existing;
    setenv(key, value.c_str(), /*overwrite=*/1);

#ifdef __APPLE__
    // Needed for DYLD_INSERT_LIBRARIES to work across two-level namespace libs.
    setenv("DYLD_FORCE_FLAT_NAMESPACE", "1", 1);
#endif
}

// ---------------------------------------------------------------------------
// Usage
// ---------------------------------------------------------------------------
static void print_usage(const char * prog) {
    printf(
        "vlm-op-profiler %s\n"
        "\n"
        "Collect ggml tensor-operation statistics from llama.cpp while running\n"
        "a vision-language model.  Outputs trace.jsonl under --out-dir.\n"
        "\n"
        "Usage:\n"
        "  %s [profiler-options] -- [llama-mtmd-cli options]\n"
        "  %s [profiler-options] --model <model.gguf> [--image <img>] [prompt]\n"
        "\n"
        "Profiler options (consumed here, not forwarded):\n"
        "  --out-dir  <path>    Output directory (default: ./profstats_out)\n"
        "  --steps    <N>       Stop recording after N decode graphs (0=unlimited)\n"
        "  --mtmd-cli <path>    Explicit path to llama-mtmd-cli\n"
        "  --include-vision-encode\n"
        "                       Also record the image-encoder graph separately\n"
        "  -h, --help           Print this message\n"
        "  --version            Print version\n"
        "\n"
        "All other flags are forwarded to llama-mtmd-cli unchanged.\n"
        "\n"
        "Environment variables (also settable here as flags):\n"
        "  PROFSTATS_OUT_DIR    Same as --out-dir\n"
        "  PROFSTATS_MAX_STEPS  Same as --steps\n"
        "\n"
        "Examples:\n"
        "  # VLM with image:\n"
        "  %s --out-dir results/test \\\n"
        "       --model  models/llava-1.6.Q4_K_M.gguf \\\n"
        "       --mmproj models/llava-1.6-mmproj.gguf \\\n"
        "       --image  photo.jpg \\\n"
        "       'Describe this image.'\n"
        "\n"
        "  # Text-only (no image):\n"
        "  %s --out-dir results/text \\\n"
        "       --model models/mistral-7b.Q4_K_M.gguf \\\n"
        "       -p 'The capital of France is'\n",
        VERSION, prog, prog, prog, prog);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
int main(int argc, char ** argv) {
    // ---- Parse profiler-specific flags ------------------------------------
    std::string out_dir          = "profstats_out";
    std::string mtmd_cli_path;
    std::size_t max_steps        = 0;
    bool        include_vision   = false;
    bool        help             = false;
    bool        show_version     = false;

    // Collect args that are NOT profiler-specific (forwarded to llama-mtmd-cli).
    std::vector<const char *> forward_args;

    for (int i = 1; i < argc; ++i) {
        const char * a = argv[i];

        if (strcmp(a, "-h") == 0 || strcmp(a, "--help") == 0) {
            help = true;
        } else if (strcmp(a, "--version") == 0) {
            show_version = true;
        } else if (strcmp(a, "--out-dir") == 0 && i + 1 < argc) {
            out_dir = argv[++i];
        } else if (strcmp(a, "--steps") == 0 && i + 1 < argc) {
            max_steps = static_cast<std::size_t>(std::atoi(argv[++i]));
        } else if (strcmp(a, "--mtmd-cli") == 0 && i + 1 < argc) {
            mtmd_cli_path = argv[++i];
        } else if (strcmp(a, "--include-vision-encode") == 0) {
            include_vision = true;  // TODO Phase 4
        } else if (strcmp(a, "--") == 0) {
            // Everything after '--' goes straight to llama-mtmd-cli.
            for (int j = i + 1; j < argc; ++j) {
                forward_args.push_back(argv[j]);
            }
            break;
        } else {
            forward_args.push_back(a);
        }
    }

    if (help) { print_usage(argv[0]); return 0; }
    if (show_version) { printf("vlm-op-profiler %s\n", VERSION); return 0; }

    if (forward_args.empty()) {
        // No model args: just show usage.
        print_usage(argv[0]);
        return 0;
    }

    // ---- Locate binaries and library -------------------------------------
    const std::string dir = self_dir();

    // libbackend_stats: must be alongside this binary.
#ifdef __APPLE__
    const std::string lib_name = "libbackend_stats.dylib";
#else
    const std::string lib_name = "libbackend_stats.so";
#endif
    const std::string lib_path = dir + lib_name;
    if (!file_exists(lib_path)) {
        fprintf(stderr,
            "vlm-op-profiler: cannot find %s (looked in %s).\n"
            "Ensure make docker-build has completed.\n",
            lib_name.c_str(), dir.c_str());
        return 1;
    }

    // llama-mtmd-cli: explicit flag, alongside binary, /app/, or PATH.
    if (mtmd_cli_path.empty()) {
        mtmd_cli_path = find_binary("llama-mtmd-cli", {
            dir + "llama-mtmd-cli",
            "/app/llama-mtmd-cli",
            dir + "bin/llama-mtmd-cli",
        });
    }
    if (mtmd_cli_path.empty()) {
        fprintf(stderr,
            "vlm-op-profiler: cannot find llama-mtmd-cli.\n"
            "Pass --mtmd-cli <path> or ensure it is on PATH.\n");
        return 1;
    }

    // ---- Create output directory -----------------------------------------
    if (!mkdirp(out_dir)) return 1;

    // ---- Configure the interceptor via environment -----------------------
    setenv("PROFSTATS_OUT_DIR", out_dir.c_str(), 1);
    {
        char buf[32];
        snprintf(buf, sizeof(buf), "%zu", max_steps);
        setenv("PROFSTATS_MAX_STEPS", buf, 1);
    }
    (void)include_vision;  // Phase 4: PROFSTATS_INCLUDE_VISION_ENCODE

    // Inject our interceptor library into the child process.
    prepend_preload(lib_path);

    // ---- Build exec argument list ----------------------------------------
    std::vector<char *> exec_argv;
    exec_argv.push_back(const_cast<char *>(mtmd_cli_path.c_str()));
    for (const char * a : forward_args) {
        exec_argv.push_back(const_cast<char *>(a));
    }
    exec_argv.push_back(nullptr);

    fprintf(stderr,
            "vlm-op-profiler: exec %s  (out_dir=%s  steps=%zu)\n",
            mtmd_cli_path.c_str(), out_dir.c_str(), max_steps);

    // exec replaces this process; on success it does not return.
    execv(mtmd_cli_path.c_str(), exec_argv.data());

    // Only reached on error.
    fprintf(stderr, "vlm-op-profiler: execv(%s): %s\n",
            mtmd_cli_path.c_str(), strerror(errno));
    return 1;
}

File: vlm_op_profiler/cli/vlm_op_profiler.py
#!/usr/bin/env python3
"""vlm-op-profiler — thin launcher that injects libbackend_stats and exec's llama-mtmd-cli.

What it does:
  1. Parses profiler-specific flags (--out-dir, --steps, --include-vision-encode, ...).
  2. Locates libbackend_stats.so/.dylib relative to this script.
  3. Creates the output directory.
  4. Writes run_meta.json (provenance: commits, model/image SHA-256, prompt, ...).
  5. Sets LD_PRELOAD / DYLD_INSERT_LIBRARIES so libbackend_stats is injected.
  6. Sets PROFSTATS_OUT_DIR / PROFSTATS_MAX_STEPS / PROFSTATS_INCLUDE_VISION_ENCODE.
  7. exec's llama-mtmd-cli with the remaining (model / image / prompt) args.

Profiler-specific flags are consumed here; everything else is forwarded to
llama-mtmd-cli unchanged.
"""

import argparse
import datetime as _dt
import hashlib
import json
import os
import platform
import shutil
import socket
import subprocess
import sys

VERSION = "0.1.0-dev"

# Directory that contains this script — used to find libbackend_stats and
# llama-mtmd-cli when they are co-located (e.g. inside the Docker image).
HERE = os.path.dirname(os.path.realpath(__file__))


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def find_binary(name: str, candidates: list[str]) -> str:
    """Return the first executable path from candidates, then search PATH."""
    for c in candidates:
        if os.path.isfile(c) and os.access(c, os.X_OK):
            return c
    return shutil.which(name) or ""


def prepend_preload(lib_path: str) -> None:
    """Prepend lib_path to LD_PRELOAD (or DYLD_INSERT_LIBRARIES on macOS)."""
    if sys.platform == "darwin":
        key = "DYLD_INSERT_LIBRARIES"
        # Needed for DYLD_INSERT_LIBRARIES to work across two-level namespace libs.
        os.environ["DYLD_FORCE_FLAT_NAMESPACE"] = "1"
    else:
        key = "LD_PRELOAD"
    existing = os.environ.get(key, "")
    os.environ[key] = lib_path + (":" + existing if existing else "")


def sha256_file(path: str) -> str | None:
    """Compute SHA-256 of a file; return None if path is missing/unreadable."""
    if not path or not os.path.isfile(path):
        return None
    try:
        h = hashlib.sha256()
        with open(path, "rb") as f:
            for chunk in iter(lambda: f.read(1 << 20), b""):
                h.update(chunk)
        return h.hexdigest()
    except OSError:
        return None


def extract_arg(argv: list[str], names: tuple[str, ...]) -> str | None:
    """Find the value of the first of `names` that appears as a flag in argv.

    Supports both `--flag value` and `--flag=value` forms. Does not consume
    or rewrite argv; this is for metadata extraction only.
    """
    for i, tok in enumerate(argv):
        if tok in names and i + 1 < len(argv):
            return argv[i + 1]
        for name in names:
            prefix = name + "="
            if tok.startswith(prefix):
                return tok[len(prefix):]
    return None


def git_commit(repo_dir: str) -> str | None:
    """Return the HEAD commit SHA of repo_dir, or None if unavailable."""
    if not os.path.isdir(repo_dir):
        return None
    try:
        out = subprocess.run(
            ["git", "-C", repo_dir, "rev-parse", "HEAD"],
            check=True,
            capture_output=True,
            text=True,
            timeout=5,
        )
        return out.stdout.strip() or None
    except (FileNotFoundError, subprocess.SubprocessError):
        return None


def resolve_profiler_commit() -> str:
    """Profiler commit: prefer baked-in env (Docker), fall back to local git."""
    env = os.environ.get("PROFILER_COMMIT")
    if env and env != "unknown":
        return env
    # When run from a source checkout, walk up from this file to find a .git.
    cur = HERE
    for _ in range(6):
        if os.path.isdir(os.path.join(cur, ".git")):
            sha = git_commit(cur)
            if sha:
                return sha
        cur = os.path.dirname(cur)
    return env or "unknown"


def resolve_llama_commit() -> str:
    """llama.cpp commit: prefer baked-in env (Docker), fall back to submodule."""
    env = os.environ.get("LLAMA_COMMIT")
    if env and env != "unknown":
        return env
    # Look for the submodule relative to this file, then walk up.
    cur = HERE
    for _ in range(6):
        cand = os.path.join(cur, "third_party", "llama.cpp")
        if os.path.exists(os.path.join(cand, ".git")) or os.path.isdir(
            os.path.join(cand, ".git")
        ):
            sha = git_commit(cand)
            if sha:
                return sha
        cur = os.path.dirname(cur)
    return env or "unknown"


def detect_inner_backend() -> str:
    """Best-effort guess at which ggml backend will be used at runtime."""
    if sys.platform == "darwin":
        return "metal"
    if os.environ.get("CUDA_VISIBLE_DEVICES") or shutil.which("nvidia-smi"):
        return "cuda"
    return "cpu"


def platform_label() -> str:
    if sys.platform == "darwin":
        return "macOS"
    if sys.platform.startswith("linux"):
        return "Linux"
    return sys.platform


def write_run_meta(
    out_dir: str,
    forward_argv: list[str],
    include_vision_encode: bool,
) -> None:
    """Write run_meta.json into out_dir.

    Fields are best-effort; missing values become null. See docs/output_format.md.
    """
    model_path = extract_arg(forward_argv, ("--model", "-m"))
    mmproj_path = extract_arg(forward_argv, ("--mmproj",))
    image_path = extract_arg(forward_argv, ("--image",))
    prompt = extract_arg(forward_argv, ("--prompt", "-p"))

    meta = {
        "profiler_version": VERSION,
        "profiler_commit": resolve_profiler_commit(),
        "llama_commit": resolve_llama_commit(),
        "model_path": model_path,
        "model_sha256": sha256_file(model_path) if model_path else None,
        "mmproj_path": mmproj_path,
        "mmproj_sha256": sha256_file(mmproj_path) if mmproj_path else None,
        "gguf_metadata": {},  # reserved for later — see docs/output_format.md
        "prompt": prompt,
        "image_path": image_path,
        "image_sha256": sha256_file(image_path) if image_path else None,
        "run_id": _dt.datetime.now(_dt.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "host": socket.gethostname(),
        "platform": platform_label(),
        "arch": platform.machine(),
        "inner_backend": detect_inner_backend(),
        "include_vision_encode": include_vision_encode,
    }

    path = os.path.join(out_dir, "run_meta.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(meta, f, indent=2, sort_keys=True)
        f.write("\n")
    print(f"vlm-op-profiler: wrote {path}", file=sys.stderr)


# ---------------------------------------------------------------------------
# Argument parser
# ---------------------------------------------------------------------------

def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(
        prog="vlm-op-profiler",
        description=(
            f"vlm-op-profiler {VERSION}\n\n"
            "Collect ggml tensor-operation statistics from llama.cpp while running\n"
            "a vision-language model.  Outputs trace.jsonl and run_meta.json under --out-dir."
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "All other flags are forwarded to llama-mtmd-cli unchanged.\n\n"
            "Environment variables (also settable here as flags):\n"
            "  PROFSTATS_OUT_DIR                Same as --out-dir\n"
            "  PROFSTATS_MAX_STEPS              Same as --steps\n"
            "  PROFSTATS_INCLUDE_VISION_ENCODE  Same as --include-vision-encode\n\n"
            "Examples:\n"
            "  # VLM with image:\n"
            "  vlm-op-profiler --out-dir results/test \\\n"
            "       --include-vision-encode \\\n"
            "       --model  models/llava-1.6.Q4_K_M.gguf \\\n"
            "       --mmproj models/llava-1.6-mmproj.gguf \\\n"
            "       --image  photo.jpg \\\n"
            "       'Describe this image.'\n\n"
            "  # Text-only (no image):\n"
            "  vlm-op-profiler --out-dir results/text \\\n"
            "       --model models/mistral-7b.Q4_K_M.gguf \\\n"
            "       -p 'The capital of France is'"
        ),
    )
    p.add_argument(
        "--version", action="version", version=f"vlm-op-profiler {VERSION}"
    )
    p.add_argument(
        "--out-dir",
        default="profstats_out",
        help="Output directory (default: ./profstats_out)",
    )
    p.add_argument(
        "--steps",
        type=int,
        default=0,
        help="Stop recording after N decode graphs (0=unlimited)",
    )
    p.add_argument(
        "--mtmd-cli",
        default="",
        metavar="PATH",
        help="Explicit path to llama-mtmd-cli",
    )
    p.add_argument(
        "--include-vision-encode",
        action="store_true",
        help=(
            "Also record vision-encoder graphs (identified by CONV_2D nodes). "
            "Default skips them so trace.jsonl reflects the LLM body only."
        ),
    )
    return p


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = build_parser()

    # Split argv on '--': everything after it is forwarded verbatim to
    # llama-mtmd-cli and never seen by argparse.
    argv = sys.argv[1:]
    try:
        sep = argv.index("--")
        our_argv, forward_argv = argv[:sep], argv[sep + 1:]
    except ValueError:
        our_argv, forward_argv = argv, None

    args, extra = parser.parse_known_args(our_argv)

    # When there was no '--', unknown args from parse_known_args are forwarded.
    if forward_argv is None:
        forward_argv = extra

    if not forward_argv:
        parser.print_help()
        sys.exit(0)

    # ---- Locate shared library -----------------------------------------------
    lib_name = (
        "libbackend_stats.dylib" if sys.platform == "darwin"
        else "libbackend_stats.so"
    )
    lib_path = os.path.join(HERE, lib_name)
    if not os.path.exists(lib_path):
        print(
            f"vlm-op-profiler: cannot find {lib_name} (looked in {HERE}).\n"
            "Ensure make docker-build has completed.",
            file=sys.stderr,
        )
        sys.exit(1)

    # ---- Locate llama-mtmd-cli -----------------------------------------------
    mtmd_cli = args.mtmd_cli or find_binary("llama-mtmd-cli", [
        os.path.join(HERE, "llama-mtmd-cli"),
        "/app/llama-mtmd-cli",
        os.path.join(HERE, "bin", "llama-mtmd-cli"),
    ])
    if not mtmd_cli:
        print(
            "vlm-op-profiler: cannot find llama-mtmd-cli.\n"
            "Pass --mtmd-cli <path> or ensure it is on PATH.",
            file=sys.stderr,
        )
        sys.exit(1)

    # ---- Create output directory and write run metadata ----------------------
    os.makedirs(args.out_dir, exist_ok=True)
    write_run_meta(args.out_dir, forward_argv, args.include_vision_encode)

    # ---- Configure the interceptor via environment --------------------------
    os.environ["PROFSTATS_OUT_DIR"] = args.out_dir
    os.environ["PROFSTATS_MAX_STEPS"] = str(args.steps)
    os.environ["PROFSTATS_INCLUDE_VISION_ENCODE"] = (
        "1" if args.include_vision_encode else "0"
    )

    prepend_preload(lib_path)

    # ---- exec into llama-mtmd-cli -------------------------------------------
    exec_argv = [mtmd_cli] + forward_argv
    print(
        f"vlm-op-profiler: exec {mtmd_cli}"
        f"  (out_dir={args.out_dir}  steps={args.steps}"
        f"  include_vision_encode={args.include_vision_encode})",
        file=sys.stderr,
    )
    os.execv(mtmd_cli, exec_argv)


if __name__ == "__main__":
    main()
